pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,348 @@
+			ID:   instanceID,
+			Tags: tags,
+		}
+		p.logger.WithFields(logrus.Fields{
+			"instance": instanceID,
+			"ip":       lookupIP,
+			"tags":     tags,
+		}).Debug("Added tags")
+	}
+	return instanceIPs, returnErr
+}
+
+// MaxInstancesBatch returns maximum number of instances that could be requested via the Instance method.
+func (p *Provider) MaxInstancesBatch() int {
+	// This is arbitrary since we have a local cache of information
+	return 64
+}
+
+// Name returns the name of the provider.
+func (p *Provider) Name() string {
+	return ProviderName
+}
+
+// SelfIP returns host's IPv4 address.
+func (p *Provider) SelfIP() (gostatsd.IP, error) {
+	// This IP is only used for start/stop events of gostatsd. To simplify the k8s provider we have
+	// chosen to just avoid finding the IP.
+	return gostatsd.UnknownIP, nil
+}
+
+// NewProviderFromViper returns a new k8s provider.
+func NewProviderFromViper(v *viper.Viper, logger logrus.FieldLogger, version string) (gostatsd.CloudProvider, error) {
+	k := util.GetSubViper(v, "k8s")
+	setViperDefaults(v, version)
+
+	// Set up the k8s client
+	clientset, err := createKubernetesClient(k.GetString(ParamUserAgent), k.GetString(ParamKubeconfigPath),
+		k.GetString(ParamKubeconfigContext), k.GetFloat64(ParamAPIQPS), k.GetFloat64(ParamAPIQPSBurstFactor))
+	if err != nil {
+		return nil, err
+	}
+
+	return NewProvider(
+		logger,
+		clientset,
+		PodInformerOptions{
+			ResyncPeriod: k.GetDuration(ParamResyncPeriod),
+			WatchCluster: k.GetBool(ParamWatchCluster),
+			NodeName:     k.GetString(ParamNodeName),
+		},
+		k.GetString(ParamAnnotationTagRegex),
+		k.GetString(ParamLabelTagRegex))
+}
+
+// NewProvider returns a new k8s provider.
+func NewProvider(logger logrus.FieldLogger, clientset kubernetes.Interface, podInfOpts PodInformerOptions,
+	annotationRegex, labelRegex string) (gostatsd.CloudProvider, error) {
+
+	// Set up the pod informer which fills an index with the pods we care about
+	indexers := cache.Indexers{
+		PodsByIPIndexName: podByIpIndexFunc,
+	}
+	customWatchOptions := func(*meta_v1.ListOptions) {}
+
+	// If we're not watching the entire cluster we need to limit our watch to pods with our node name
+	if !podInfOpts.WatchCluster {
+		if podInfOpts.NodeName == "" {
+			return nil, fmt.Errorf("watch-cluster set to false, and node name not supplied")
+		}
+		customWatchOptions = func(lo *meta_v1.ListOptions) {
+			lo.FieldSelector = fmt.Sprintf("spec.nodeName=%s", podInfOpts.NodeName)
+		}
+	}
+
+	podsInf := core_v1inf.NewFilteredPodInformer(
+		clientset,
+		meta_v1.NamespaceAll,
+		podInfOpts.ResyncPeriod,
+		indexers,
+		customWatchOptions)
+
+	// TODO: we should emit prometheus metrics to fit in with the k8s ecosystem
+	// TODO: we should emit events to the k8s API to fit in with the k8s ecosystem
+
+	return &Provider{
+		logger:          logger,
+		podsInf:         podsInf,
+		annotationRegex: regexp.MustCompile(annotationRegex),
+		labelRegex:      regexp.MustCompile(labelRegex),
```

Comment:
I'd change this method to accept the regex type, not strings. "Accept direct dependencies" is the "rule" of dependency inversion/injection. It's the job of the caller to construct them.